Fix to work on NetBSD/powerpc

he32 opened the pull request:
This pull request adds the necessary bits to make ocaml work
on NetBSD/powerpc again.  I've verified that after these fixes,
"make test" completes with no unexpected failures.

he32 committed:
Add missing glue bits to support NetBSD/powerpc.

he32 committed:
Change NetBSD/powerpc from elf to bsd_elf.

The rest of the configure setup appears to make the (arguably invalid)
assumption that powerpc,elf => Linux.

he32 committed:
Change test to check NATDYNLINK instead of SUPPORTS_SHARED_LIBRARIES.

NetBSD/powerpc does not yet do "native dynamic linking", and this
test would otherwise fail.

he32 committed:
Note fix for NetBSD/powerpc.

lthls committed:
Fix duplication of code in Cmmgen (#1370)

* Fix duplicates in Cmmgen when handling switches with no default and not all cases

* Improve handling of incomplete Lambda switches in Flambda

* Add test (for reference) and changes

* Fix nitpick

* Cleanup: split the switch stores to reflect usage

* Improve compilation of incomplete switches with flambda

* Split switch stores into context-aware and -unaware versions

* Credit reviewers

* Go back to a single Switch.t_store type

lthls committed:
Fix duplication of code in Cmmgen (#1370)

* Fix duplicates in Cmmgen when handling switches with no default and not all cases

* Improve handling of incomplete Lambda switches in Flambda

* Add test (for reference) and changes

* Fix nitpick

* Cleanup: split the switch stores to reflect usage

* Improve compilation of incomplete switches with flambda

* Split switch stores into context-aware and -unaware versions

* Credit reviewers

* Go back to a single Switch.t_store type

gasche commented on asmrun/signals_osdep.h:
```diff
@@ -287,7 +307,7 @@
 /****************** PowerPC, ELF (Linux) */
 
-#elif defined(TARGET_power) && defined(SYS_elf)
+#elif defined(TARGET_power) && defined(SYS_elf) && !defined(__NetBSD__)
```
Why do you need this if you set `system=bsd_elf` in configure? Does a system configured this way still define `SYS_elf`? (I'm not familiar with this part of the configuration logic, and would have assumed that only `SYS_bsd_self` was defined).

he32 replied:
This is a remnant from when I tried to use the existing setting `system=elf`, I'll clean this up.

gasche commented on configure:
```diff
@@ -973,7 +973,7 @@ case "$target" in
                                 if $arch64; then model=ppc64; else model=ppc; fi
                                 system=elf;;
-  powerpc-*-netbsd*)            arch=power; model=ppc; system=elf;;
+  powerpc-*-netbsd*)            arch=power; model=ppc; system=bsd_elf;;
```
This was set to `elf` (originally it was `bsd`) to fix [MPR#4116](https://caml.inria.fr/mantis/view.php?id=4116). Any idea of why the setting needs to change now?

he32 replied:
Because with `system=elf`, we fall into the
    `power,gcc*,elf,ppc)   partialld="ld -r -m elf32ppclinux";;`
part which we don't want to.

gasche commented on testsuite/tests/lib-dynlink-native/Makefile:
```diff
@@ -23,7 +23,7 @@ LD_PATH = $(TOPDIR)/otherlibs/$(UNIXLIBVAR)unix:$(TOPDIR)/otherlibs/systhreads\
 .PHONY: default
 default:
-	@if ! $(SUPPORTS_SHARED_LIBRARIES) || $(BYTECODE_ONLY) ; then \
+	@if ! $(NATDYNLINK) || $(BYTECODE_ONLY) ; then \
```
This looks correct: the test uses `Dynlink` directly in [main.ml](https://github.com/he32/ocaml/blob/e5fbdb17a70b8222b04f37f3defaa33c5a71980b/testsuite/tests/lib-dynlink-native/main.ml) so I would assume that it cannot work unless `NATDYNLINK` is set.

gasche commented on asmrun/signals_osdep.h:
```diff
@@ -154,6 +154,26 @@
  #define CONTEXT_YOUNG_PTR (context->uc_mcontext.gregs[REG_R15])
  #define CONTEXT_FAULTING_ADDRESS ((char *) info->si_addr)
 
+/****************** PowerPC, NetBSD */
```
The blocks in this file seem to be ordered by architecture first, OS second, so I think it would make more sense to have this blob in the PowerPC part below. Currently the "PowerPC, BSD" check is clearly designed to also catch NetBSD systems (except it doesn't as the NetBSD configure uses `system=elf`), so it would make sense to have the new code close by, and also maybe to disable the `SYS_netbsd` conditional there?

he32 replied:
Ok, I'll look at moving this block as appropriate.  The use of `system=elf` doesn't actually work, because "power,elf" seems to imply Linux, as configure then wants to do `"partialld="ld -r -m elf32ppclinux"`.  I have a new change lined up which will instead use `system=netbsd` also for NetBSD/powerpc, I'm just waiting for all the self-tests to complete.

gasche reviewed (commented):
I don't know anything about this part of the codebase, so this is not an informed review, but I had a look and asked questions.

he32 commented:
Hm, on second thoughts, an additional diff to configure is coming up.
The earlier set of changes worked with 4.05.00.

he32 committed:
Move the NetBSD/powerpc section to the appropriate place.

Also undo the "don't do SYS_elf for NetBSD", as we will now use
SYS_netbsd for NetBSD/powerpc.

he32 committed:
Convert to use system=netbsd for NetBSD/powerpc.

he32 committed:
Change to use SYS_netbsd for NetBSD/powerpc.

Also remove the later SYS_netbsd test which is now inactive.

he32 committed:
Move the s390 section so the power sections are together.

he32 commented:
And with this I now have the test summary, run on NetBSD/macppc 8.0_BETA:
```Summary:
  748 tests passed
   23 tests skipped
    0 tests failed
    0 unexpected errors
  771 tests considered
```

gasche commented:
Would you mind rebasing your patchset to hide the back-and-forth on `system=` setting?

he32 commented:
Hm, rebasing is stretching my git-abilities (as is evident since a change from trunk snuck in here).
Let me see if I can delete and re-create the branch instead.
So ... I'll probably abandon this pull request and create another one.

he32 committed:
Add a section specific to NetBSD/powerpc.

he32 committed:
Move the s390 section so the powerpc sections are together.

he32 committed:
Change NetBSD/powerpc to use system=netbsd.

This avoids us falling into partialld="ld -r -m elf32ppclinux"
which will not work.
Also fix pattern for other powerpc/bsd* (typo fix).

he32 committed:
Change test to check NATDYNLINK instead of SUPPORTS_SHARED_LIBRARIES.

he32 committed:
Note fix for NetBSD/powerpc.

he32 committed:
Merge branch 'netbsd-ppc' of https://github.com/he32/ocaml into netbsd-ppc
